### DISCUSSION TITLE
Use jsPsych.timelineVariable not as parameter but in a function. Possible?

### DISCUSSION THREAD

**Discussion by DanielBruijn** [jspsych/jsPsych#1755] [Q&A] [answered]
Hello all! I am very stuck and much obliged to anyone that can help. 

I've defined a full-factorial design, and i want to use the values in there in my stimulus displays. Normally, you'd use something like 'stimulus: jsPsych.timelineVariable('stimulus')'. However, my stimulus is a rather complex function, it needs to draw specific shapes on specific points and all that, so that function is defined beforehand (called drawStim). 

The factorial design:
```
factors = {
  eccentricity: [0,1],
  set_size: [1,4],
  delay: [150, 250, 500],
};

full_design = jsPsych.randomization.factorial(factors, 32);
half_design = jsPsych.randomization.factorial(factors, 2);

```
Now my question is, is it possible to use the values from within the factorial design in that function? Since in the 'normal' way, timelineVariable is called as a parameter, and it is not a plain value like a number.

I tried putting a definition in the function, like this: (lines 5 and 6)
````
function drawStim(canvas) {
  var ctx = canvas.getContext('2d');
  var center_y = Math.floor(canvas.height / 2);
  var center_x = Math.floor(canvas.width / 2);
  eccent = jsPsych.timelineVariable('eccentricity');//shuffle(Array(0,1))[0]; //determines eccentricity at set size 1
  set_size = jsPsych.timelineVariable('set_size');
 (...)
  if(set_size == 1) { 
    one_pos = stim_pos[0]; 
    locations = one_pos //for data saving
    
    ctx.beginPath();
    if(eccent == 1) { //outer eccentricity
      ctx.arc(x_coords[one_pos], y_coords[one_pos], 40, 0, 2 * Math.PI);
    } else if(eccent == 0) { //inner eccentricity
      ctx.arc(x_coords_small[one_pos], y_coords_small[one_pos], 40, 0, 2 * Math.PI);
    };
    ctx.fill();
  } else if(set_size == 4) {

(...)  a whole lot more stuff in the function (...) 

  // Fixation cross
  ctx.fillStyle = 'black';
  ctx.font = '60px san-serif';
  ctx.textAlign = 'center';
  ctx.textBaseline = 'middle';
  ctx.fillText("+", center_x, center_y);
}
````

But that doesn't work, it can't recognize the timelineVariable, i guess because it isn't a number. When i display the object eccent for example, it shows 'function() { return timeline.timelineVariable(varname); }.

Can anyone help me out to properly use those values from the design in my function?

**Comment by becky-gilbert** [comment, marked as answer]
Hi @DanielBruijn, I think you need to add the second `true` argument to `jsPsych.timelineVariable`, which tells jsPsych to return the value immediately, rather than returning a function that returns the value. For example, inside your `drawStim` function you should be able to access the current value of timeline variables like this:
```js
eccent = jsPsych.timelineVariable('eccentricity', true);
```
This `true` argument shouldn't be necessary if you're using jsPsych 6.3.0+.

**Reply by DanielBruijn** [reply]
Thanks for your suggestion! I tried that, unfortunately it didn't work like i wanted to. I ended up writing separate functions for the eccentricity and set size conditions, so just four functions, and call those via the timelineVariable. This works perfectly!

Also, it might have to do with the fact that i'm running 6.2.0 instead, i had some issues with 6.3.0. 

Thanks again for your time, if i have some spare time i'll try to make a second version that works well via this route!
